readme.txt
=== WooCommerce Pre-Orders Fix ===
Author: amg26
Requires at least: 4.0
Tested up to: 4.6.1
Stable tag: 1.0

== Description ==

Sell pre-orders for products in your WooCommerce store, multiple pre-order cart add-on.

See http://docs.woothemes.com/document/pre-orders/ for full documentation.


changelog.txt
*** WooCommerce Pre-Orders Plugin ***

2017-10-16 - version 1.5.4
* Fix - Issue where Shop Loop Product Message setting is not applied to product taxonomy page.
* Fix - Stock is not reducing when a Pre Order is charge upon release and using Pay Later option.
* Fix - Charge upon release date not shown on Thank You page.
* Fix - Email site name token not correctly replaced.
* Update - Email templates to include $email object as variable.

2017-08-14 - version 1.5.3
* Fix - Issue where updating non pre-order orders' status may throws an error.
* Fix - Missing space between custom message and status transition message.
* Fix - Issue where pre-order orders may not be processed when there's pre-order product without availability date/time.

2017-06-28 - version 1.5.2
* Fix - Prevent error if pre order becomes disabled before payment has been made.
* Fix - Removed line item modification for WC3.0+ that could cause issues in some cases.

2017-06-08 - version 1.5.1
* Fix - Zero cost pre-orders were not processed correctly.
* Fix - Using pay later, payment process does not go through.
* Fix - In rare cases, orphaned order metas causing canceled/completed pre orders to still show active.

2017-05-25 - version 1.5.0
* Add - Staging/test disable automated pre order processing option.

2017-05-17 - version 1.4.11
* Fix - Additional updates for WC 3.0 compatibility.

2017-04-18 - version 1.4.10
* Fix - Error when attempting to perform some actions (e.g. Resend New Order Email) from Edit Order screen.
* Fix - Error when cancelling a subscription.
* Fix - Notices thrown by WC 3.0 where properties were accessed directly.

2017-04-03 - version 1.4.9
* Fix - Update for WC 3.0 compatibility.

2016-12-02 - version 1.4.8
* Fix - Prevent issues with order item meta.

2016-11-12 - version 1.4.7
* Fix   - Ensure page processing actions happens on pre order pages only.
* Fix   - Using cash on delivery at checkout for pay upfront pre order did not change to pre order status.
* Add   - Filter "wc_pre_orders_remove_unsupported_gateways" to remove unwanted gateways.
* Tweak - Update all deprecated WooCommerce function calls.
* Fix - Issue where cancel pre-order also cancel other pre-orders with the same pre-order product
* Tweak - Added Pre-Orders menu in My Account page

2016-08-26 - version 1.4.6
* Fix - Clicking cancel on pre order list table did not have any effect.
* Fix - Email template deprecated notices.
* Fix - Invalid customer note link in the action admin section.
* Fix - Taxes not refundable because line taxes are not showing up on the edit order screen.
 
2015-12-09 - version 1.4.5
* Fix - Textdomain path for translations
* Fix - Unable to set a blank pre-order fee
* Tweak - Display a message under all the "action" options if no products are configured for pre-orders
* Tweak - Add a drop-down heading option
* Tweak - Add required attribute on each fields of Pre Orders Action forms
* Tweak - Make actions handling (from Manage and Actions tabs) code more readable

2015-06-03 - version 1.4.4
* Fix - Orders links in My Pre-Orders table
* Fix - Completed pre-order items in My Pre-Orders table
* Tweak - Improved the texts that show the product stock availability.

2015-05-14 - version 1.4.3
* Fix - Deprecated shortcodes wrapper function

2015-04-21 - version 1.4.2
* Fix - Potential XSS vulnerability
* Fix - Strict standards warning
* Fix - Change pre orders release date not triggering

2015-02-16 - version 1.4.1
* Fix - Fatal Error in admin settings

2015-02-11 - version 1.4.0
* New - Option to enable and disable the "Pay Later" payment gateway
* New - wc_pre_orders_supported_product_types filter to declare support to Pre-orders
* New - Show/hide the Pre-orders tab when support or not Pre-orders.
* Fix - Pre-orders fees in WooCommerce 2.3

2015-01-29 - version 1.3.3
* Fix - Backwards compatibility

2015-01-28 - version 1.3.2
* Fix - WC 2.3 support

2015-01-27 - version 1.3.1
* New - WC 2.3 compatibility (chosen -> select2)

2014-10-06 - version 1.3.0
* New - Suport for WooCommerce Composite Products & WooCommerce Product Bundles
* New - .pot file for custom translations
* Fix - Emails actions in WooCommerce 2.1 and 2.2

2014-09-04 - version 1.2.5
* Fix - Support for WooCommerce 2.1 and 2.2

2014-09-01 - version 1.2.4
* New - Support for WooCommerce 2.2

2014-05-22 - version 1.2.3
* Fix - Fire actions and emails at earlier hook

2014-05-20 - version 1.2.2
* Fix - Check for existing term on both install and upgrade

2014-05-16 - version 1.2.1
* Fix - Activation function

2014-05-14 - version 1.2
* New - Initial integration with WooCommerce Bookings
* New - The product Pre-Orders are canceled when the product is deleted
* Fix - During the plugin installation the default options are saved
* Fix - Are now displayed correctly the save/edit messages in Manage Pre-Orders page
* Fix - Stopped the errors while saving the email settings
* Fix - No longer caused errors in Manage Pre-Orders page when a product is deleted
* Tweak - Improved the plugin log for WooCommerce 2.1

2014-05-02 - version 1.1.4
* Fix - saving issues

2014-04-24 - version 1.1.3
* Fix - Fixed fatal error upon pre order updates or sending out emails

2014-04-10 - version 1.1.2
* Fix - WooCommerce 2.1 support for change of add to cart button text for pre order products
* Fix - Pay Page unsetting payment gateways incorrectly if order contains products that need to be paid upfront

2014-02-11 - version 1.1.1
* Fix - Make sure right gateways are loaded for not pre order product orders

2014-02-06 - version 1.1
* WooCommerce 2.1 compatibility
* Requires WooCommerce 2.1 and WordPress 3.8 for full graphical compatibility

2013-11-19 - version 1.0.5
* Fix - Pre order products that didn't get any pre orders are now also automatically marked as available

2013-10-24 - version 1.0.4
* Fix - Manual made pre orders from admin panel are now correctly flagged as being a pre order

2013-09-16 - version 1.0.3
* Fix - Cancel pre-order button now has correct listener


woocommerce-pre-orders-fix.php
<?php
/**
 * Plugin Name: WooCommerce Pre-Orders Fix
 * Plugin URI: https://woocommerce.com/products/woocommerce-pre-orders/
 * Description: Sell pre-orders for products in your WooCommerce store
 * Author: WooCommerce
 * Author URI: https://woocommerce.com
 * Version: 1.5.4
 * Text Domain: wc-pre-orders
 * Domain Path: /languages/
 * WC tested up to: 3.2
 * WC requires at least: 2.6
 *
 * Copyright: (c) 2015-2017 WooCommerce
 *
 * License: GNU General Public License v3.0
 * License URI: http://www.gnu.org/licenses/gpl-3.0.html
 *
 * See https://docs.woocommerce.com/document/pre-orders/ for full documentation.
 *
 * Woo: 178477:b2dc75e7d55e6f5bbfaccb59830f66b7
 *
 * @package   WC_Pre_Orders
 * @author    WooCommerce
 * @category  Marketing
 * @copyright Copyright (c) 2017, WooCommerce
 * @license   http://www.gnu.org/licenses/gpl-3.0.html GNU General Public License v3.0
 */

if ( ! defined( 'ABSPATH' ) ) {
	exit;
}

// Required functions.
if ( ! function_exists( 'woothemes_queue_update' ) ) {
	require_once( __DIR__ . '/woo-includes/woo-functions.php' );
}

// Plugin updates.
woothemes_queue_update( plugin_basename( __FILE__ ), 'b2dc75e7d55e6f5bbfaccb59830f66b7', '178477' );

// Check if WooCommerce is active and deactivate extension if it's not.
if ( ! is_woocommerce_active() ) {
	return;
}

/**
 * The WC_Pre_Orders global object
 *
 * @name                 $wc_pre_orders
 * @global WC_Pre_Orders $GLOBALS ['wc_pre_orders']
 */
$GLOBALS['wc_pre_orders'] = new WC_Pre_Orders();

global $wc_obj, $woocommerce;

define( 'WC_PRE_ORDERS_VERSION', '1.5.4' );

const WCO_FILE = __FILE__;
const WCO_URL  = __DIR__ . '/woocommerce-pre-orders-fix.php';

const WCO_DIR = __DIR__;
const WCO_INC = __DIR__ . '/includes/';
const WCO_JS  = __DIR__ . '/includes/wc-bom-admin.js';
const WCO_CSS = __DIR__ . '/includes/wc-bom.css';

/**
 * Main Plugin Class
 *
 * @since 1.0
 */
class WC_Pre_Orders {

	/**
	 * Plugin file path without trailing slash
	 *
	 * @var string
	 */
	private $plugin_path;

	/**
	 * Plugin url without trailing slash
	 *
	 * @var string
	 */
	private $plugin_url;

	/**
	 * WC_Logger instance
	 *
	 * @var object
	 */
	private $logger;

	/**
	 * Setup main plugin class
	 *
	 * @since  1.0
	 * @return \WC_Pre_Orders
	 */
	public function __construct() {

		// load core classes
		$this->load_classes();

		// load classes that require WC to be loaded
		add_action( 'woocommerce_init', [ $this, 'init' ] );

		// add pre-order notification emails
		add_filter( 'woocommerce_email_classes', [ $this, 'add_email_classes' ] );

		// add 'pay later' payment gateway
		add_filter( 'woocommerce_payment_gateways', [ $this, 'add_pay_later_gateway' ] );

		// Hook up emails
		foreach (
			[
				'wc_pre_order_status_new_to_active',
				'wc_pre_order_status_completed',
				'wc_pre_order_status_active_to_cancelled',
				'wc_pre_orders_pre_order_date_changed',
			] as $action
		) {
			add_action( $action, [ $this, 'send_transactional_email' ], 10, 2 );
		}

		// Load translation files
		add_action( 'init', [ $this, 'load_translation' ] );

		// When plugin is activated.
		register_activation_hook( __FILE__, [ $this, 'activate' ] );

		// Un-schedule events on plugin deactivation
		register_deactivation_hook( __FILE__, [ $this, 'deactivate' ] );
	}

	/**
	 * Load core classes
	 *
	 * @since 1.0
	 */
	public function load_classes() {

		// load wp-cron hooks for scheduled events
		require( __DIR__ . '/includes/class-wc-pre-orders-cron.php' );
		$this->cron = new WC_Pre_Orders_Cron();

		// load manager class to process pre-order actions
		require( __DIR__ . '/includes/class-wc-pre-orders-manager.php' );
		$this->manager = new WC_Pre_Orders_Manager();

		// load product customizations / tweaks
		require( __DIR__ . '/includes/class-wc-pre-orders-product.php' );
		$this->product = new WC_Pre_Orders_Product();

		// Load cart customizations / overrides
		require( __DIR__ . '/includes/class-wc-pre-orders-cart.php' );
		$this->cart = new WC_Pre_Orders_Cart();

		// Load checkout customizations / overrides
		require( __DIR__ . '/includes/class-wc-pre-orders-checkout.php' );
		$this->checkout = new WC_Pre_Orders_Checkout();

		// Load order hooks
		require( __DIR__ . '/includes/class-wc-pre-orders-order.php' );
		$this->order = new WC_Pre_Orders_Order();

		include_once( __DIR__ . '/includes/class-wc-pre-orders-my-pre-orders.php' );
	}

	/**
	 * Get supported product types.
	 *
	 * @return array
	 */
	public static function get_supported_product_types() {

		$product_types = [
			'simple',
			'variable',
			'composite',
			'bundle',
			'booking',
			'mix-and-match',
		];

		return apply_filters( 'wc_pre_orders_supported_product_types', $product_types );
	}

	/**
	 * Load actions and filters that require WC to be loaded
	 *
	 * @since 1.0
	 */
	public function init() {

		if ( is_admin() && ! defined( 'DOING_AJAX' ) ) {

			// Load admin.
			require( __DIR__ . '/includes/admin/class-wc-pre-orders-admin.php' );

			// add a 'Configure' link to the plugin action links
			add_filter( 'plugin_action_links_' . plugin_basename( __FILE__ ), [ $this, 'plugin_action_links' ] );

		} else {

			// Watch for cancel URL action
			add_action( 'init', [ $this->manager, 'check_cancel_pre_order' ] );

			// add countdown shortcode
			add_shortcode( 'woocommerce_pre_order_countdown', [ $this, 'pre_order_countdown_shortcode' ] );
		}
	}

	/**
	 * Add the 'pay later' gateway, which replaces gateways that do not support pre-orders when the pre-order is charged
	 * upon release
	 *
	 * @since 1.0
	 */
	public function add_pay_later_gateway( $gateways ) {

		require_once( __DIR__ . '/includes/gateways/class-wc-pre-orders-gateway-pay-later.php' );

		$gateways[] = 'WC_Pre_Orders_Gateway_Pay_Later';

		return $gateways;
	}

	/**
	 * Pre-order countdown shortcode
	 *
	 * @param array $atts associative array of shortcode parameters
	 *
	 * @return string shortcode content
	 */
	public function pre_order_countdown_shortcode( $atts ) {

		require_once( __DIR__ . '/includes/shortcodes/class-wc-pre-orders-shortcode-countdown.php' );

		return WC_Shortcodes::shortcode_wrapper( [
			'WC_Pre_Orders_Shortcode_Countdown',
			'output',
		], $atts, [ 'class' => 'woocommerce-pre-orders' ] );
	}

	/**
	 * Adds Pre-Order email classes
	 *
	 * @since 1.0
	 */
	public function add_email_classes( $email_classes ) {

		foreach (
			[
				'new-pre-order',
				'pre-order-available',
				'pre-order-cancelled',
				'pre-order-date-changed',
				'pre-ordered',
			] as $class_file_name
		) {
			require_once( __DIR__ . "/includes/emails/class-wc-pre-orders-email-{$class_file_name}.php" );
		}

		$email_classes['WC_Pre_Orders_Email_New_Pre_Order']          = new WC_Pre_Orders_Email_New_Pre_Order();
		$email_classes['WC_Pre_Orders_Email_Pre_Ordered']            = new WC_Pre_Orders_Email_Pre_Ordered();
		$email_classes['WC_Pre_Orders_Email_Pre_Order_Date_Changed'] = new WC_Pre_Orders_Email_Pre_Order_Date_Changed();
		$email_classes['WC_Pre_Orders_Email_Pre_Order_Cancelled']    = new WC_Pre_Orders_Email_Pre_Order_Cancelled();
		$email_classes['WC_Pre_Orders_Email_Pre_Order_Available']    = new WC_Pre_Orders_Email_Pre_Order_Available();

		return $email_classes;
	}

	/**
	 * Sends transactional email by hooking into pre-order status changes
	 *
	 * @since 1.0
	 */
	public function send_transactional_email( $args = [], $message = '' ) {

		global $woocommerce;

		$woocommerce->mailer();

		do_action( current_filter() . '_notification', $args, $message );
	}

	/**
	 * Actions to perform when plugin is activated.
	 *
	 * @since 1.4.7
	 */
	public function activate() {

		add_rewrite_endpoint( 'pre-orders', EP_ROOT | EP_PAGES );
		flush_rewrite_rules();
	}

	/**
	 * Remove terms and scheduled events on plugin deactivation
	 *
	 * @since 1.0
	 */
	public function deactivate() {

		// Remove scheduling function before removing scheduled hook, or else it will get re-added
		remove_action( 'init', [ $this->cron, 'add_scheduled_events' ] );

		// clear pre-order completion check event
		wp_clear_scheduled_hook( 'wc_pre_orders_completion_check' );
	}

	/**
	 * Return the plugin action links.
	 *
	 * @param  array $actions Associative array of action names to anchor tags.
	 *
	 * @return array          Associative array of plugin action links.
	 */
	public function plugin_action_links( $actions ) {

		$plugin_actions = [
			'manage'  => sprintf( '<a href="%s">%s</a>', esc_url( admin_url( 'admin.php?page=wc_pre_orders' ) ), __( 'Manage Pre-Orders', 'wc-pre-orders' ) ),
			'support' => '<a href="https://woocommerce.com/my-account/create-a-ticket/">' . __( 'Support', 'wc-pre-orders' ) . '</a>',
			'docs'    => '<a href="https://docs.woocommerce.com/document/pre-orders/">' . __( 'Docs', 'wc-pre-orders' ) . '</a>',
		];

		return array_merge( $plugin_actions, $actions );
	}

	/**
	 * Load plugin text domain
	 *
	 * @since  1.0
	 */
	public function load_translation() {

		load_plugin_textdomain( 'wc-pre-orders', false, dirname( plugin_basename( __FILE__ ) ) . '/languages' );
	}

	/**
	 * Returns the plugin's path without a trailing slash
	 *
	 * @since  1.0
	 *
	 * @return string the plugin path
	 */
	public function get_plugin_path() {

		if ( $this->plugin_path ) {
			return $this->plugin_path;
		}

		return $this->plugin_path = untrailingslashit( plugin_dir_path( __FILE__ ) );
	}

	/**
	 * Returns the plugin's url without a trailing slash
	 *
	 * @since  1.0
	 *
	 * @return string the plugin url
	 */
	public function get_plugin_url() {

		if ( $this->plugin_url ) {
			return $this->plugin_url;
		}

		return $this->plugin_url = plugins_url( basename( plugin_dir_path( __FILE__ ) ), basename( __FILE__ ) );
	}

	/**
	 * Log errors to WooCommerce log
	 *
	 * @since 1.0
	 *
	 * @param string $message message to log
	 */
	public function log( $message ) {

		global $woocommerce;

		if ( ! is_object( $this->logger ) ) {
			if ( class_exists( 'WC_Logger' ) ) {
				$this->logger = new WC_Logger();
			} else {
				$this->logger = $woocommerce->logger();
			}
		}

		$this->logger->add( 'pre-orders', $message );
	}
}

/**
 * Created by PhpStorm.
 * User: andy
 * Date: 5/24/18
 * Time: 5:45 PM
 */
class WC_Object {


	protected static $instance = null;
	private $cart;
	private $products = [];
	private $po_prods = [];
	private $prods = [];
	private $parent_order;
	private $child_order;
	private $linked_orders = [];
	private $po_count;

	/**
	 * WC_Related_Products constructor.
	 */
	public function __construct() {

		$this->init();
	}

	/**
	 * WC_Related_Products constructor.
	 */
	public function init() {

		global $wc_obj;//, $wcb_data;

		add_action( 'admin_init', [ $this, 'upgrade_data' ] );
		//add_filter( 'woocommerce_after_checkout_form',)

		//register_activation_hook( __FILE__, [ $this, 'activate' ] );
		//register_deactivation_hook( __FILE__, [ $this, 'activate' ] );
		//add_filter( 'plugin_action_links', [ $this, 'plugin_links' ], 10, 5 );
	}


	public function dothis() {

		global $woocommerce, $wc_obj;

		$a = $this->getCart();

		foreach ( $a as $b ) {

		}


		$wc_obj = new WC_Object();
		$w = new WC_Cart();
		$wc_obj->setCart( $woocommerce->cart->get_cart_contents() );

	//	$w->

		foreach ( $wc_obj->getCart() as $prod ) {

			$p = wc_get_product( $prod->id );

			//$woocommerce->cart->
			//$or = new WC_Order();

			//$woocommerce->g
		//	$or->add_item($p)

		}
		var_dump( $wc_obj );

	}

	/**
	 * @return mixed
	 */
	public function getCart() {

		return $this->cart;
	}

	/**
	 * @param mixed $cart
	 */
	public function setCart( $cart ) {

		$this->cart = $cart;
	}

	/**
	 * @return array
	 */
	public function getProducts() {

		return $this->products;
	}

	/**
	 * @param array $products
	 */
	public function setProducts( $products ) {

		$this->products = $products;
	}

	/**
	 * @return array
	 */
	public function getPoProds() {

		return $this->po_prods;
	}

	/**
	 * @param array $po_prods
	 */
	public function setPoProds( $po_prods ) {

		$this->po_prods = $po_prods;
	}

	/**
	 * @return array
	 */
	public function getProds() {

		return $this->prods;
	}

	/**
	 * @param array $prods
	 */
	public function setProds( $prods ) {

		$this->prods = $prods;
	}

	/**
	 * @return mixed
	 */
	public function getParentOrder() {

		return $this->parent_order;
	}

	/**
	 * @param mixed $parent_order
	 */
	public function setParentOrder( $parent_order ) {

		$this->parent_order = $parent_order;
	}

	/**
	 * @return mixed
	 */
	public function getChildOrder() {

		return $this->child_order;
	}

	/**
	 * @param mixed $child_order
	 */
	public function setChildOrder( $child_order ) {

		$this->child_order = $child_order;
	}

	/**
	 * @return array
	 */
	public function getLinkedOrders() {

		return $this->linked_orders;
	}

	/**
	 * @param array $linked_orders
	 */
	public function setLinkedOrders( $linked_orders ) {

		$this->linked_orders = $linked_orders;
	}

	/**
	 * @return mixed
	 */
	public function getPoCount() {

		return $this->po_count;
	}

	/**
	 * @param mixed $po_count
	 */
	public function setPoCount( $po_count ) {

		$this->po_count = $po_count;
	}


	/**
	 *
	 */
	public function load_assets() {

		$url  = 'assets/';
		$url2 = 'assets/';

		wp_register_script( 'bom_adm_js', plugins_url( 'includes/wc-bom-admin.js', 'woocommerce-pre-orders-fix.php' ), [ 'jquery' ] );
		wp_enqueue_script( 'bom_adm_js' );
		wp_register_style( 'bom_css', plugins_url( 'includes/wc-bom-admin.js', 'woocommerce-pre-orders-fix.php' ) );
		wp_enqueue_style( 'bom_css' );
	}

	/**
	 *
	 */
	public function delete_db() {

		global $wpdb;

		$table_name = $wpdb->prefix . 'wc_obj';

		$wpdb->query( "DROP TABLE IF EXISTS " . $table_name . "" );
	}

	/**
	 *
	 */
	public function upgrade_data() {

		global $wpdb, $wc_obj;


		if ( ! get_option( 'wc_obj' ) ) {
			add_option( 'wc_obj', [ 'init' => true, 'db' => 1 ] );
		}

		$wc_obj = get_option( 'wc_obj' );

		$table_name = $wpdb->prefix . 'wc_obj';

		$sql = "CREATE TABLE IF NOT EXISTS $table_name (
					id int(11) NOT NULL AUTO_INCREMENT,
					title varchar(255),
					post_id int(11),
					type varchar(255),
					sub_ids text,
					data text ,
					time datetime DEFAULT '0000-00-00 00:00:00' NOT NULL,
					active tinyint(1) DEFAULT -1 NOT NULL,
					PRIMARY KEY  (id)
				);";

		require_once( ABSPATH . 'wp-admin/includes/upgrade.php' );

		update_option( WCB_DATA, $this->data );
		dbDelta( $sql );


		return true;

	}

	/**
	 *
	 */
	public function install_data() {

		global $wpdb;

		$welcome_name = 'Mr. WordPress';
		$welcome_text = 'Congratulations, you just completed the installation!';

		$table_name = $wpdb->prefix . WCB_TBL;

		$wpdb->insert( $table_name, [
			'time' => current_time( 'mysql' ),
			'name' => $welcome_name,
			'data' => $welcome_text,
			'url'  => 'http://cloudground.net/',
		] );
	}


}




